p4-16/spec/P4-16-spec.mdk

```diff
@@ -1180,7 +1180,113 @@ produced by different programmers at different times:
+instead the users should employ the newer module `import` directive
+described in section [#sec-modules].
+
+## Modules { #sec-modules }
+
+P4 offers a very simple module system.  Modules enable programmers to
+decompose a program into multiple source files and libraries.
+
+In the compilation process the preprocessing step is executed first.
+After preprocessing the modules are loaded.  Modules offer a
+functionality similar to `#include` directives:
+
+* Both enable a program to be decomposed into multiple source files.
+
+* Both need to some degree to understand the filesystem layout of
+  modules.  The standard C preprocessor does this through the use of
+  the `-I` flag to specify the location of directories where included
+  files have to searched.  We recommend the use of the `-L` flag for a
+  similar role for specifying filesystem paths to be searched by the
+  P4 compiler for `import`-ed modules.  In general, the compiler will
+  maintain an ordered list of module library paths, and every `import`
+  statement will search the specified files in order under each of the
+  library paths, loading the first file that matches.
+
+There are also some important differences between modules and
+`#included` files:
+
+* `#include` directives produce flat files.  In contrast, P4 modules
+  can optionally create new namespaces, allowing P4 programs to make
+  use of multiple libraries without concern about name clashes.
+
+* `#include` directives do not understand anything about the P4 syntax
+  and semantics, they are purely string processing operations.  In
+  particular, the included files do not have to be syntactically
+  correct P4 programs or program fragments.
+
+* `#include` directives function transitively: a file included from
+  another included file has the exact same visibility as the top-level
+  constructs.  In contrast, transitively included modules are
+  completely invisible at the top-level.
+
+The following keywords are used to construct modules in P4:
+
+* `import`, `into`: import definitions from a module
+* `private`: mark definitions of a module as non-importable
+
+### Importing modules { #sec-import }
+
+A definitions in a module can be imported into the global namespace
+with the following statement, executed at compile-time.
+
+`import module/path/file.p4`
+
+~ Begin P4Grammar
+importStatement: IMPORT filePath
+               | IMPORT filePath INTO namespacePath
+               ;
+
+filePath: FILENAME
+        | DIRECTORYNAME SLASH filePath
+        ;
+
+namespacePath: IDENTIFIER
+             | namespacePath DOT IDENTIFIER
+             ;
+~ End P4Grammar
+
+`import` statements must be present at the top-level only.  The effect
+of the `import` statement is to include all the definitions in the
+imported module in the place of the `import` module.  An import is
+followed by an optional path separated by slashes and by a file name.
+Only *relative* paths are allowed.
```

Comment:
relative to what?  What exactly does an imported module inherit from the context where it is imported?  Command line arguments?  Anything else?

How do nested imports work.  If I have `import A int X` and `A` contains `import B into Y`, does `B` end up in `Y` or `X.Y`?

Reply:
The path is relative in that it cannot start with /.
It will be relative to the specified paths (probably using command-line arguments).

Nested imports are not visible outside the importer, so the problem does not arise at all.